### PR TITLE
fix: guarding gas config in useTransactionCallback

### PR DESCRIPTION
**Pull request by UncleBill**
## Description

Fix the problem that requires input gas in MetaMask manually.

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/mask/src/plugins/MaskBox/hooks/useContext.ts [hunk 1]
```diff
@@ -253,8 +253,8 @@ function useContext(initialState?: { boxId: string }) {
     const canPurchase = !isBalanceInsufficient && isQualified && !!boxInfo?.personalRemaining
     const allowToPurchase = boxState === BoxState.READY
     const isAllowanceEnough = isNativeToken ? true : costAmount.lte(erc20Allowance ?? '0')
-    const { value: openBoxTransactionGasLimit = 0 } = useAsyncRetry(async () => {
-        if (!openBoxTransaction || !canPurchase || !allowToPurchase || !isAllowanceEnough) return 0
+    const { value: openBoxTransactionGasLimit } = useAsyncRetry(async () => {
+        if (!openBoxTransaction || !canPurchase || !allowToPurchase || !isAllowanceEnough) return
```
Comment (UncleBill) [inline packages/mask/src/plugins/MaskBox/hooks/useContext.ts:257]: Unset, let next step make the decision.

**Review by zhouhanseng — DISMISSED** [review]

File: packages/mask/src/plugins/MaskBox/helpers/formatCountdown.ts [hunk 2]
```diff
@@ -13,20 +13,16 @@ const units = [
...
+    const segs: string[] = []
 
-    ;[msPerDay, msPerHour, msPerMinute, msPerSecond].reduce((accumulator, x) => {
-        amounts.push(Math.floor(accumulator / x))
+    ;[msPerDay, msPerHour, msPerMinute, msPerSecond].reduce((accumulator, x, index, list) => {
+        const isLast = index === list.length - 1
+        const value = isLast ? Math.ceil(accumulator / x) : Math.floor(accumulator / x)
+        if (segs.length !== 0 || value !== 0) {
+            segs.push(`${value} ${units[index][value === 1 ? 0 : 1]}`)
+        }
```
Comment (UncleBill) [inline packages/mask/src/plugins/MaskBox/helpers/formatCountdown.ts:23]: Show `1 hour 0 minute 0 second` instead of `1 hour`

Comment (septs) [reply packages/mask/src/plugins/MaskBox/helpers/formatCountdown.ts]: why not use https://date-fns.org/v2.27.0/docs/intervalToDuration?

Comment (septs) [reply packages/mask/src/plugins/MaskBox/helpers/formatCountdown.ts]: e.g:
```
formatDuration(intervalToDuration({
    start: new Date(0),
    end: new Date(countdown * 1000)
}))
```

File: packages/mask/src/plugins/MaskBox/helpers/formatCountdown.ts [hunk 3]
```diff
@@ -1,32 +1,7 @@
...
-    return amounts
-        .map((x, i) => {
-            if (x <= 0) return ''
-            if (x === 1) return `${x} ${units[i][0]}`
-            return `${x} ${units[i][1]}`
-        })
-        .filter(Boolean)
-        .join(' ')
-        .trim()
+export function formatCountdown(from: number, to: number) {
```
Comment (septs) [inline packages/mask/src/plugins/MaskBox/helpers/formatCountdown.ts:4]:
```suggestion
export function formatCountdown(from: Date | number, to: Date | number = Date.now()) {
```

File: packages/web3-shared/evm/hooks/useTransactionCallback.ts [hunk 4]
```diff
@@ -21,9 +21,13 @@ export function useTransactionCallback<T extends unknown>(
         setState({
             type: TransactionStateType.WAIT_FOR_CONFIRMING,
         })
+        const gasExpectedConfig = { ...config }
```
Comment (guanbinrui) [inline packages/web3-shared/evm/hooks/useTransactionCallback.ts:24]: Do the web3.js alter this?

File: packages/mask/src/plugins/MaskBox/hooks/useContext.ts [hunk 5]
```diff
@@ -135,17 +137,18 @@ function useContext(initialState?: { boxId: string }) {
 
     const boxState = useMemo(() => {
         if (errorMaskBoxInfo || errorMaskBoxStatus || errorBoxInfo) return BoxState.ERROR
-        if (loadingMaskBoxInfo || loadingMaskBoxStatus || loadingBoxInfo) return BoxState.UNKNOWN
+        if (loadingMaskBoxInfo || loadingMaskBoxStatus || loadingBoxInfo) {
+            if (!maskBoxInfo && !boxInfo) return BoxState.UNKNOWN
```
Comment (UncleBill) [inline packages/mask/src/plugins/MaskBox/hooks/useContext.ts:141]: When we already know something,  don't lie.

When call retry callback, UI won't flicker.

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**